### PR TITLE
configure: add --disable-windows-unicode option

### PR DESCRIPTION
**Pull request by nojb**
Adds a `configure` option to disable Unicode support under Windows. See [MPR#7904](https://caml.inria.fr/mantis/view.php?id=7904).

### CODE REVIEW

**Comment by avsm** [comment]
As a CI maintainer who is slightly grumpy about the number of configure-time options we have to test in opam, could I gently ask what the usecase/impact for this toggle? Will this need to be tested on all submitted opam packages when we activate Windows CI with both runtime options?  Or is the flag available for advanced users only who know what they are doing wrt Windows unicode?

**Comment by nojb** [comment]
> As a CI maintainer who is slightly grumpy about the number of configure-time options we have to test in opam, could I gently ask what the usecase/impact for this toggle?

The usecase is that Unicode support is not backwards compatible (think about strings that were stored in a DB in Latin-1 encoding which are retrieved and passed to functions now expecting UTF-8). In such environments, migration to Unicode mode needs to be carefully managed.

> Will this need to be tested on all submitted opam packages when we activate Windows CI with both runtime options?

No, it doesn't need to be tested.

> Or is the flag available for advanced users only who know what they are doing wrt Windows unicode?

Only for advanced users.

**Comment by nojb** [comment]
On second thoughts, I don't think it is a good idea to have this as an option in `configure`. If one needs this, better to edit the `Makefile` directly.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/25 21:58 -0800):
> On second thoughts, I don't think it is a good idea to have this as an
> option in `configure`. If one needs this, better to edit the
> `Makefile` directly.

Well really I don't like the idea of having to edit the makefiles and
find the configure option a much better interface.

@nojb would you really mind reopening this?

**Comment by nojb** [comment]
> @nojb would you really mind reopening this?

OK, sure.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/25 22:51 -0800):
> > @nojb would you really mind reopening this?
> 
> OK, sure.

Thanks! I'll review it today. To give a concrete argument in favour of
this PR, I think having a configure option makes it easier to provide
compiler packages where Unicode support for Windows is disabled. Of
course not having the option does not make this strictly impossible, but
still less straightforward.

Will review this today.

**Comment by shindere** [comment]
The code looks correct to me. I have just one question regarding the
help text. Currently it says: "disable Windows Unicode runtime support".
How about "disable Unicode support in Windows runtime" ?

**Comment by dra27** [comment]
I completely agree with @shindere that there should be nothing which requires editing the generated files from `configure` - it's highly unhygienic. There's also no need to test the Unicode stuff in CI - as @nojb points out, it is an advanced option for legacy support.

**Comment by avsm** [comment]
Many thanks for the clarifications!

**Comment by dra27** [comment]
For the implementation, there are several modes available here, and it would be better if we now supported them all properly. Unicode mode is *always* enabled, because we use the wide functions, so from a sort of hygiene-perspective, I don't think it should be referred to as disabling Unicode.

I would do this using `AC_ARG_VAR` perhaps called `WINDOWS_UNICODE_MODE` (`WINDOWS_UNICODE_TRANSLATION_MODE` is a bit too much of a mouthful?!) to reflect that, which would have a default value of `compatible` and should accept values of:

 - `ansi` - causes `windows_unicode_enabled = 0` in `runtime/win32.c` (i.e. sets `WINDOWS_UNICODE` to `0`)
 - `nonstrict` - `windows_unicode_enabled = 1` and `windows_unicode_strict = 0` (**this is new**)
 - `compatible` - `windows_unicode_enabled = 1` and `windows_unicode_fallback = 1` (that's the current default)
 - `strict` - `windows_unicode_enabled = 1`, `windows_unicode_strict = 1` and `windows_unicode_fallback = 0` (**this is new**)

**Comment by dra27** [comment]
Incidentally, at this point, I'd have no issue with only the `ansi` and `compatible` values being implemented - i.e. we put the infrastructure in place to expose the other flags later, but don't require any build system alterations at this time.

**Comment by dra27** [comment]
NB For @shindere, my comment above previously incorrect suggested a default of `fallback` when it should have been a default of `compatible`

**Comment by shindere** [comment]
Thanks a lot for your comment, @dra27 !

**Comment by shindere** [comment]
David Allsopp (2019/02/26 09:59 +0000):
> Incidentally, at this point, I'd have no issue with only the `ansi`
> and `compatible` values being implemented - i.e. we put the
> infrastructure in place to expose the other flags later, but don't
> require any build system alterations at this time.

That looks to me as a nice and smooth way to proceed!

**Comment by shindere** [comment]
David Allsopp (2019/02/26 02:01 -0800):
> NB For @shindere, my comment above previously incorrect suggested a
> default of `fallback` when it should have been a default of
> `compatible`

Ah thanks! I guess you editted the comment and you mention it here so
that I get a mail, right?

**Comment by nojb** [comment]
> For the implementation, there are several modes available here, and it would be better if we now supported them all properly. Unicode mode is _always_ enabled, because we use the wide functions, so from a sort of hygiene-perspective, I don't think it should be referred to as disabling Unicode.
> 
> I would do this using `AC_ARG_VAR` perhaps called `WINDOWS_UNICODE_MODE` (`WINDOWS_UNICODE_TRANSLATION_MODE` is a bit too much of a mouthful?!) to reflect that, which would have a default value of `compatible` and should accept values of:
> 
> * `ansi` - causes `windows_unicode_enabled = 0` in `runtime/win32.c` (i.e. sets `WINDOWS_UNICODE` to `0`)
> * `nonstrict` - `windows_unicode_enabled = 1` and `windows_unicode_strict = 0` (**this is new**)
> * `compatible` - `windows_unicode_enabled = 1` and `windows_unicode_fallback = 1` (that's the current default)
> * `strict` - `windows_unicode_enabled = 1`, `windows_unicode_strict = 1` and `windows_unicode_fallback = 0` (**this is new**)

Wish granted. @shindere do you mind taking another look?

**Review by dra27 — APPROVED** [review]
Lovely, thank you!

**Comment by dra27** [comment]
This strikes me as both safe and a good idea for 4.08.0 - @gasche, @damiendoligez ?

**Comment by gasche** [comment]
Yes, I think this can go in 4.08.

**Comment by nojb** [comment]
Does it need a CHANGES entry?

**Comment by gasche** [comment]
I think it's a good idea to have a Changes entry, to let other Windows users know that it exists.

**Comment by nojb** [comment]
> I think it's a good idea to have a Changes entry, to let other Windows users know that it exists.

OK, I added one.

**Comment by damiendoligez** [comment]
Yes, this can go into 4.08.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/26 14:10 +0000):
> > For the implementation, there are several modes available here, and it would be better if we now supported them all properly. Unicode mode is _always_ enabled, because we use the wide functions, so from a sort of hygiene-perspective, I don't think it should be referred to as disabling Unicode.
> > 
> > I would do this using `AC_ARG_VAR` perhaps called `WINDOWS_UNICODE_MODE` (`WINDOWS_UNICODE_TRANSLATION_MODE` is a bit too much of a mouthful?!) to reflect that, which would have a default value of `compatible` and should accept values of:
> > 
> > * `ansi` - causes `windows_unicode_enabled = 0` in `runtime/win32.c` (i.e. sets `WINDOWS_UNICODE` to `0`)
> > * `nonstrict` - `windows_unicode_enabled = 1` and `windows_unicode_strict = 0` (**this is new**)
> > * `compatible` - `windows_unicode_enabled = 1` and `windows_unicode_fallback = 1` (that's the current default)
> > * `strict` - `windows_unicode_enabled = 1`, `windows_unicode_strict = 1` and `windows_unicode_fallback = 0` (**this is new**)
> 
> Wish granted. @shindere do you mind taking another look?

Absolutely not!

I just would like to nitpick about "according to the value of the
environment variable +  WINDOWS_UNICODE_MODE" in Changes. Would it be
possible to use a wording that suggests that this should be used as an
argument to configure rather than defined as an environment variable?

The rest looks fine to me.

**Comment by dra27** [comment]
It's a little bike-sheddy, but cf. https://github.com/ocaml/ocaml/pull/2266#issuecomment-467581437 regarding the `Changes` entry ... updating the documentation might be more worthwhile (not sure which bit...!)

**Comment by nojb** [comment]
I updated the `Changes` entry to remove the word "environment".

**Comment by nojb** [comment]
> It's a little bike-sheddy, but cf. [#2266 (comment)](https://github.com/ocaml/ocaml/pull/2266#issuecomment-467581437) regarding the `Changes` entry ... updating the documentation might be more worthwhile (not sure which bit...!)

If I understand, it is "OK" to leave this Changes entry separate or should I put it together with the rest of #1200?

Is there a section in the manual that discusses the options to `configure`?

**Comment by nojb** [comment]
Also, this is to be squashed before merging.

**Comment by dra27** [comment]
@nojb - my suggestion was that the GPR number could simply be added to GPR#2139 (the autoconf PR) on the basis that it's simply an addition to that work.

**Comment by nojb** [comment]
> @nojb - my suggestion was that the GPR number could simply be added to GPR#2139 (the autoconf PR) on the basis that it's simply an addition to that work.

OK, will do.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/27 09:00 +0000):
> > @nojb - my suggestion was that the GPR number could simply be added to GPR#2139 (the autoconf PR) on the basis that it's simply an addition to that work.
> 
> OK, will do.

I'd prefere not, actually.

I really like to have separate entries for GPRs when their authors or
reviewers are distinct. To me, having a single entry in such cases tends
to be not only non acucrate but also misleading.

**Comment by nojb** [comment]
> Nicolás Ojeda Bär (2019/02/27 09:00 +0000):
> > @nojb - my suggestion was that the GPR number could simply be added to GPR#2139 (the autoconf PR) on the basis that it's simply an addition to that work. OK, will do.
> I'd prefere not, actually. I really like to have separate entries for GPRs when their authors or reviewers are distinct. To me, having a single entry in such cases tends to be not only non acucrate but also misleading.

OK, in that case this PR is ready to merge. I will squash, merge and cherry-pick shortly unless someone complains.

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/27 01:52 -0800):
> OK, in that case this PR is ready to merge. I will squash, merge and
> cherry-pick shortly unless someone complains.

Fine with me, go ahead!

**Comment by nojb** [comment]
Cherry-picked to 4.08 as daf9f27c9e1d9023f0957b304b71e707a37256e0

**Comment by shindere** [comment]
Nicolás Ojeda Bär (2019/02/26 23:01 -0800):
> I updated the `Changes` entry to remove the word "environment".

Thanks a lot for having done so, @nojb!

**Pull request merged**
